dockerclient/client.go

```diff
@@ -609,6 +610,53 @@ func (e *ClientExecutor) Preserve(path string) error {
 	if e.Volumes == nil {
 		e.Volumes = NewContainerVolumeTracker()
 	}
+	createPath := func(dest string) error {
+		var writerErr error
+		if !strings.HasSuffix(dest, "/") {
+			dest = dest + "/"
+		}
+		reader, writer := io.Pipe()
+		opts := docker.UploadToContainerOptions{
+			InputStream: reader,
+			Path:        "/",
+			Context:     context.TODO(),
+		}
+		go func() {
+			defer writer.Close()
+			tarball := tar.NewWriter(writer)
+			defer tarball.Close()
+			writerErr = tarball.WriteHeader(&tar.Header{
+				Name:     dest,
+				Typeflag: tar.TypeDir,
+				Mode:     0755,
+			})
+		}()
+		glog.V(4).Infof("Uploading empty archive to %q", dest)
+		err := e.Client.UploadToContainer(e.Container.ID, opts)
+		if err != nil {
+			return fmt.Errorf("unable to ensure existence of preserved path %s: %v", dest, err)
+		}
+		if writerErr != nil {
+			return fmt.Errorf("error generating tarball to ensure existence of preserved path %s: %v", dest, writerErr)
+		}
+		return nil
+	}
+	readPath := func(dest string) error {
+		if !strings.HasSuffix(dest, "/") {
+			dest = dest + "/"
+		}
+		err := e.Client.DownloadFromContainer(e.Container.ID, docker.DownloadFromContainerOptions{
+			Path:         dest,
+			OutputStream: ioutil.Discard,
+		})
+		return err
+	}
+	if err := readPath(path); err != nil {
```

Comment:
Why is the check necessary?  Can't we just always upload the dir and rely on later stages to fail (for instance, RUN would fail because cd would fail).

Reply:
If the volume directory already existed in the container, this would overwrite its ownership unconditionally.  I'll add a conformance test that fails if the take this bit out.